docs/RFCS/20222202_historical_hot_ranges.md

```diff
@@ -0,0 +1,460 @@
+budget.
+
+A dedicated single-tenant deployment only performs host sampling, while
+multi-tenant deployments perform both tenant and host sampling. In a
+single-tenant deployment, the results of tenant's sampling sampling would be
+contained in the host sampling, and therefore be redundant.
+
+Tenant sampling and Host sampling follow mostly the same process - the
+differences manifest in how aggregation is performed to keep samples within
+budget, and how the sampled metrics are written to `system.hot_ranges`.
+
+According to the sample rate, a job is executed to
+request the 128 hottest replicas from every store via `replRankings`. These hot
+replicas are then sorted alphabetically by their start keys. If tenant sampling
+is being performed, each set of a tenant's reported replicas is
+limited to 1000 by a range reduction and aggregation scheme. This range
+reduction scheme is discussed in the next section. If host sampling is being
+performed, the global set of reported replicas is limited to 1000 by the same
+range reduction scheme. 
+
+After ranges have been aggregated and reduced, they are written to
+`system.hot_ranges`. Tenant sampling creates a row per tenant per sample, while
+Host sampling creates a single row per sample, whose `tenant_id` value is set to
+`'host'`. To query `system.hot_ranges`, demateralized views are created for each
+tenant and for the host.
+
+Finally, existing samples older than 2 weeks will be deleted.
+
+
+
+```sql 
+-- schema for system.hot_ranges
+CREATE TABLE system.hot_ranges ( 
+    sample_time INT NOT NULL,
+    
+    -- When host sampling is being performed, 
+    -- tenant_id will be set to 'host'
+    tenant_id INT NOT NULL,
+
+    -- info is defined below by `HHRSample`
+    info BYTES NOT NULL,
+
+    PRIMARY KEY (sample_time, tenant_id)
+); 
+```
+
+```protobuf
+// protobuf definition for HHRSample
+message HHRSample { 
+    message HotRange {
+        message Schema { 
+            string database = 1;
+            repeated string tables = 2; 
+        } 
+        int32 range_id = 1; 
+        int32 replica_id = 2; 
+        float qps = 3; 
+        string start_key = 4; 
+        string end_key = 5; 
+        repeated int32 node_ids = 6;
+        repeated int32 store_ids = 7; 
+        repeated Schema schema = 8; 
+        string locality = 9; 
+        int32 bytes = 10;
+    }
+    repeated HotRange hotranges = 1;
+} 
+```
+
+
+```sql 
+-- Dematerialzied view for a tenant with tenant id, t.
+CREATE VIEW system.tenant_hot_ranges 
+AS SELECT * FROM system.hot_ranges 
+WHERE tenant_id = t;
+
+-- dematerialized view for a host
+CREATE VIEW system.host_hot_ranges 
+AS SELECT * FROM system.hot_ranges
+WHERE tenant_id = 'host';
+```
+
+
+The sampling process is pictured here: ![Metric
+Generation](./images/20222202_historical_hot_ranges/metric-generation.png
+"Metric Generation")
+
+
+### Range Reduction and Aggregation
+
+Range reduction and aggregation ensures that no more than 1000 metrics are
```

Comment:
did you consider an algorithm that considered history? By that I mean, that looked at the previous aggregation decisions and tried to balance the desire to maximize resolution of hot ranges with the desire to keep aggregations stable between pairs of entries?

This aggregation section is the section I'm most concerned about when it comes to multiple dimensions. Is the idea that we'd store the same buckets but each bucket would have multiple dimensions or would we just copy-paste the entire structure for the different dimensions? The latter being multiplicative in the size of the storage. 

I think we know quite deeply that the current metric is insufficient and we're actively fixing that. I want to make sure we don't paint ourselves in a corner here. Perhaps there's a way to sketch out an entropy metric that combines the different dimensions.

Reply:
>did you consider an algorithm that considered history?

Yeah, the aggregation scheme will evolve over time.  Using the previous sample to inform which part of the key space should have higher resolution is something I want to experiment with too.


> Is the idea that we'd store the same buckets but each bucket would have multiple dimensions or would we just copy-paste the entire structure for the different dimensions?

This is a good question. I think it depends if the user expects multiple dimensions to be reported by the same aggregate ranges. If that's the case (and I suspect it will be), then it would be the former.